arch/rv32i/src/syscall.rs

```diff
@@ -0,0 +1,389 @@
+//! Kernel-userland system call interface for RISC-V architecture.
+
+use core::fmt::Write;
+
+use kernel;
+
+/// This holds all of the state that the kernel must keep for the process when
+/// the process is not executing.
+#[derive(Copy, Clone, Default)]
+#[repr(C)]
+pub struct RiscvimacStoredState {
+    /// Store all of the app registers.
+    regs: [usize; 31],
+
+    /// This holds the PC value of the app when the exception/syscall/interrupt
+    /// occurred. We also use this to set the PC that the app should start
+    /// executing at when it is resumed/started.
+    pc: usize,
+
+    /// We need to store the mcause CSR between when the trap occurs and after
+    /// we exit the trap handler and resume the context switching code.
+    mcause: usize,
+}
+
+/// Implementation of the `UserspaceKernelBoundary` for the RISC-V architecture.
+pub struct SysCall();
+
+impl SysCall {
+    pub const unsafe fn new() -> SysCall {
```

Comment:
unnecessary `unsafe` alternatively add some docs which explain why it is marked as unsafe

Reply:
Agree that this is only meaningful if SysCall cannot otherwise be easily instatiated by anyone (which I believe this `unsafe` is meant to protect)

Reply:
I think this is either a copy-and-paste artifact, or the `unsafe` was needed at some point during the development, and the compiler only tells you when it is necessary to add it.